include/cache.h

```diff
@@ -13,78 +13,229 @@
 extern "C" {
 #endif
 
-void arch_dcache_flush(void *addr, size_t size);
-void arch_dcache_invd(void *addr, size_t size);
+/**
+ * Common operations for the caches
+ *
+ * WB means write-back and intends to transfer dirty cache lines to memory in a
+ * copy-back cache policy. May be a no-op in write-back cache policy.
+ *
+ * INVD means invalidate and will mark cache lines as not valid. A future
+ * access to the associated address is guaranteed to generate a memory fetch.
+ */
+#define K_CACHE_WB	BIT(0)
+#define K_CACHE_INVD	BIT(1)
+#define K_CACHE_WB_INVD	(K_CACHE_WB | K_CACHE_INVD)
+
+/**
+ *
+ * @brief Enable d-cache
+ *
+ * Enable the d-cache.
+ *
+ * @return N/A
+ */
+void arch_dcache_enable(void);
+
+/**
+ *
+ * @brief Disable d-cache
+ *
+ * Disable the d-cache.
+ *
+ * @return N/A
+ */
+void arch_dcache_disable(void);
+
+/**
+ *
+ * @brief Enable i-cache
+ *
+ * Enable the i-cache.
+ *
+ * @return N/A
+ */
+void arch_icache_enable(void);
+
+/**
+ *
+ * @brief Disable i-cache
+ *
+ * Disable the i-cache.
+ *
+ * @return N/A
+ */
+void arch_icache_disable(void);
+
+/**
+ *
+ * @brief Write-back / Invalidate / Write-back + Invalidate one level d-cache
+ *
+ * Write-back, Invalidate or Write-back + Invalidate one lever of the d-cache
+ *
+ * @param level	The cache level to act on
+ * @param op	Operation to perform (one of the K_CACHE_* operations)
+ *
+ * @retval 0		On success
+ * @retval -ENOTSUP	If the operation is not supported
+ */
+int arch_dcache_level(int level, int op);
```

Comment:
Hmm, I'm wondering do we really need to expand such low-level cache management API to generic code?
Any example about why generic code may want to do some operations with exact cache level?

Reply:
reading around it seems that sometimes for example you want to invalidate L1 cache on ARM v7 cores but it's really a narrow use case and for such cases it's probably better to just add a few lines of code when needed. So I agree that it's better to take this out.